src/gluonts/ev/api.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from collections import ChainMap
+from dataclasses import dataclass, field
+from typing import (
+    Callable,
+    Collection,
+    Dict,
+    Iterator,
+    Optional,
+    Protocol,
+    runtime_checkable,
+)
+
+import numpy as np
+
+from gluonts.dataset.split import TestData
+from gluonts.time_feature.seasonality import get_seasonality
+from gluonts.model.predictor import Predictor
+from .stats import seasonal_error
+from .aggregations import Aggregation
+
+
+@runtime_checkable
+class Metric(Protocol):
+    def __call__(self, axis: Optional[int] = None) -> "MetricEvaluator":
+        raise NotImplementedError
+
+
+class EvaluationDataBatch:
```

Comment:
This looks unnecessary. What benefit does this provide over a simple dict?

Reply:
I'd like to get rid of this, too, but I'm not sure how. Some explanation on why it's here:

(1) If the predictor yields single `Forecast` objects (as opposed to [`ForecastBatch`](https://github.com/awslabs/gluonts/pull/2286)es), the entries in `data` are only 1-dimensional which makes it impossible to get values per entry or per time stamp - instead, everything will be aggregated to a single number.

(2) So, the `EvaluationDataBatch` wraps everything in one more dimension to make it look as if we're dealing with a batch of size 1. While we could do this only once for the values in `non_forecast_data`, for the `forecast`s we can't (because they dynamically return values for the requested quantiles).

It is pretty wasteful to always generate new NumPy arrays when the data is accessed so it would be nice to get rid of this. If we have the `ForecastBatch` class, we could have `ForecastBatch.fromForecast(forecast)` and the chain map it with the (only once batched) `non_forecast_data` - this would remove the need for the `EvaluationDataBatch`.

Do you agree with (1) in that we need to "batchify" things when they are not given in batches or do you see another option to this?

Reply:
`EvaluationDataBatch` was removed in [this commit](https://github.com/awslabs/gluonts/pull/2267/commits/0223741cf0aea7ea8a4d45a2ef286eda65f7ce84). This change builds on top of the BatchForecast PR #2286 which is not yet merged.

Doing it like this seems pretty natural to me, though. This way, the user can work with predictors that produce `Forecast`s or `ForecastBatch`es and except for a few lines of code in the data preparation, the evaluation works the same, as was the plan.